operator/src/lib.rs

```diff
@@ -846,30 +628,71 @@ impl ControllerStrategy for SparkStrategy {
     type State = SparkState;
     type Error = error::Error;
 
-    fn error_policy(&self) -> ReconcilerAction {
-        let reconcile_after_error_sec = 30;
-        error!(
-            "Reconciliation error: requeuing after {} seconds!",
-            reconcile_after_error_sec
-        );
-        reconcile::create_requeuing_reconciler_action(Duration::from_secs(
-            reconcile_after_error_sec,
-        ))
-    }
-
     async fn init_reconcile_state(
         &self,
         context: ReconciliationContext<Self::Item>,
-    ) -> Result<Self::State, Error> {
+    ) -> Result<Self::State, Self::Error> {
+        let existing_pods = context.list_pods().await?;
+        trace!("Found [{}] pods", existing_pods.len());
+
+        let cluster_spec = &context.resource.spec;
+
+        let mut eligible_nodes = HashMap::new();
+
+        eligible_nodes.insert(
+            SparkNodeType::Master,
+            get_nodes_that_fit_selectors(&context.client, &cluster_spec.masters).await?,
+        );
+
+        eligible_nodes.insert(
+            SparkNodeType::Worker,
+            get_nodes_that_fit_selectors(&context.client, &cluster_spec.workers).await?,
+        );
+
+        if let Some(history_servers) = &cluster_spec.history_servers {
+            eligible_nodes.insert(
+                SparkNodeType::HistoryServer,
+                get_nodes_that_fit_selectors(&context.client, &history_servers).await?,
+            );
+        }
+
         Ok(SparkState {
-            spec: context.resource.spec.clone(),
-            status: context.resource.status.clone(),
             context,
-            pod_information: None,
+            existing_pods,
+            eligible_nodes,
         })
     }
 }
 
+async fn get_nodes_that_fit_selectors<T>(
```

Comment:
I can't shake the feeling that this would be better suited to be in the framework. I do something extremely similar (the same) in the Kafka Operator ...

Reply:
It is probably identical, was copied from Kafka as well!